libflux/src/core/ast/mod.rs

```diff
@@ -537,6 +538,49 @@ pub struct ArrayType {
     pub monotype: MonoType,
 }
 
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct FunctionType {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+    pub parameters: Option<Vec<ParameterType>>,
+    pub monotype: MonoType,
+}
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct ParameterType {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+    pub identifier: Option<Identifier>,
+    pub parameter: MonoType,
+    pub _type: Option<PType>,
+}
```

Comment:
Can we make this an enum instead?
```Rust
pub enum ParameterType {
    Required {
        base: BaseNode,
        name: String,
        ty: MonoType,
    },
    Optional {
        base: BaseNode,
        name: String,
        ty: MonoType,
    },
    Pipe {
        base: BaseNode,
        name: Option<String>,
        ty: MonoType,
    },
}
```
This way the type of the parameter - required, optional, or pipe - is encoded as part of the Rust type, and we don't need to keep the original tokens around.